Hello, Idan!

My name is Olga. I'm happy to reviewing your project today. 

The first time I see a mistake, I' will just point it out and let you find it and fix it yourself. In a real job, your boss will do the same, and I'm trying to prepare you to work as an Data Analyst. But if you can't handle this task yet, I will give you a more accurate hint at the next check.


Below you will find my comments - ** please do not move, modify or delete them **.

You can find my comments in green, yellow or red boxes like this:


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Success. Everything is done succesfully.
</div>


<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Remarks. Some recommendations.
</div>


<div class="alert alert-block alert-danger">
    
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
    
Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>


You can answer me by using this: 

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Overallv reviewer's comment</b> <a class="tocSkip"></a>
    
Idan, thank you for sending your project. You've done a really good job on it!

While there's room for improvement, on the whole, your project is impressive good. You did a great job! 

<span class="text-danger">There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.</span>
    
Every issue with our code is a chance for us to learn something new.

Good luck! :)
    
</div>

<div class="alert alert-block alert-warning">
<b>For your future projects</b> <a class="tocSkip"></a>
    
- Try to give tables meaningful names. Cars, clients, orders etc. df = DataFrame. Imagine, that you have 50 tables: df1, df2, df3, etc. it is hard to understand which table we need.
    
- It would be better to import all libraries in a first cell. It help us to set up of working environment.

- Using display(table name.head()) make datasets more visually attractive.

- We shoud use Kernel-Restart&Run All before sending project. It restart project and help to see errors.
</div>


<div class="alert alert-block alert-success">
<b>Overallv reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Hello Idan! I really appreciate the corrections you sent in! Thanks for taking the time to do so.
    
Excellent style!
    
Now your project is a true "A". Congratulations!
    
I'm glad to say that it has been accepted.
    
Keep up the good work, and good luck on the next sprint! :)     
    
</div>

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Great start with detailed introduction! :) </div>

## Open the data file and have a look at the general information. 

### looking at the data

In [1]:
import pandas as pd

In [2]:
try:
    df = pd.read_csv('https://code.s3.yandex.net/datasets/credit_scoring_eng.csv')
except:
    df = pd.read_csv('/datasets/real_estate_data_us.csv') 

df.info()
df.head

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<bound method NDFrame.head of        children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

         

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Well done! We have basic information about our dataset.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
It would be better to import libraries in a first cell and read datasets in a second cell. Sometimes we need to add a library. It would be easy to restart one cell that whole project. :) </div>

### looking into days_employed column

<font size='4'><center>days_employed seems weird. in addition, days_employed and total_income have the same number of missing values</center></font>
<font size = '4'><center>lets get the min/max and mean/median values for days_employed to get a grip on whats going on over there</center></font>

In [3]:
max_days_employed = df.loc[df['days_employed'].idxmax()]['days_employed']
min_days_employed = df.loc[df['days_employed'].idxmin()]['days_employed']
span_days_employed = min_days_employed,max_days_employed                           
span_days_employed

(-18388.949900568383, 401755.40047533)

<font size='4'><center>we see we have a big span of values and also that we have some negative numbers. lets check the proportion of negative numbers:</center></font>

In [4]:
negative_employ_ratio = df.loc[df['days_employed'] < 0]['days_employed'].count()/len(df)
f'{negative_employ_ratio:.1%} of days_employed values are negative'

'73.9% of days_employed values are negative'

<font size='4'><center>the data seems weird. lets take its absolute value and check the span again</center></font>

In [5]:
df['days_employed'] = df['days_employed'].abs()
max_days_employed_abs = df.loc[df['days_employed'].idxmax()]['days_employed']
min_days_employed_abs = df.loc[df['days_employed'].idxmin()]['days_employed']
span_days_employed_updated = min_days_employed_abs,max_days_employed_abs
f'updated extreme days_employed values: {span_days_employed_updated}'

'updated extreme days_employed values: (24.14163324048118, 401755.40047533)'

In [6]:
days_employed_abs_ratio = max_days_employed_abs / min_days_employed_abs
f'ratio of days_employed extreme values: {days_employed_abs_ratio}'

'ratio of days_employed extreme values: 16641.59986498587'

<font size = '4'><center>the span and ratio are so big, so we'll suggest ignoring the data: its not a problem of measuring seconds/hours/years
instead of days</center></font>

<font size = '4'><center>besides, it seems a lot of the values contain a lot of digits after the decimal dot, and its another reason to doubt the data. in the context of days_employed, its hard to think about any property connected which will have so many digits after the decimal dot</center></font>

In [7]:
mean_days = df['days_employed'].mean()
median_days = df['days_employed'].median()
print(f'mean days: {mean_days}')
print(f'median days: {median_days}')

mean days: 66914.72890682195
median days: 2194.220566878695


<font size = '4'><center>the mean is much greater than the median, which suggests a
relatively small amount of values pushes the mean upwards.
lets check the bigger values of days_employed - if theres a lot of them close to the max value, and the next big values will be around the mean, maybe we'll have a reason to delete this cluster of big 
values next to the mamx value and keep the column after all
    </center></font>

In [8]:
print(df.loc[df['days_employed'] > 5*mean_days]['days_employed'].count()/len(df))
print(df.loc[df['days_employed'] < 5*mean_days]['days_employed'].count()/len(df))
print(df['days_employed'].isna().sum()/len(df))

0.1473170731707317
0.7516840882694541
0.10099883855981417


<font size='4'><center>
15% of the values are greater than 
5 times the mean. the max_value is 6*mean_days so no cluster of points we can delete to save the column.</center></font>

### extracting relevant info from days_employed before deleting it

<font size='4'><center>before deleting it, lets check the connection between 'days_employed' missing values to the missing values of the 'total_income'. after all, they have the same number of missing values:</center></font>

print(df.loc[df['total_income'].isna()]['days_employed'].isna().sum())

<font size='4'><center>we see an exact correlation between having missing 'total_income' value to having missing 'days_employed' value. we'll presume the missing values at those rows originate from the loan seeker lack of job. maybe theyre kids that
hadnt start work yet, but somehow have a bank account and they want a loan for a summer trip. and maybe its people who are currently unemployed, so they
write their previous 'income_type' but leave the rest blank. as for businesses and civil servants - its strange they dont have a total_income
but thats the data. maybe theyre ashame about those details and prefer not to disclose them if not necessary</center></font>

<font size='4'><center>the data in 'days_employed' is irrelevant anyway. if we'll fill those missing values with 0 it wouldnt help:
it will give new information but the data we want is probably their overall aggregated work time in the last few
years or more.</center></font>

<font size ='5'><center>in conclusion, well drop the 'days employed' column. we could answer the relevant questions without it</center></font>

In [9]:
df.drop('days_employed', axis=1, inplace=True)

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Very good. We do not need this column in our results, so we can do anything. :) </div>


### fixing the education column

<font size='4'><center>lets fix the 'education' column as well. see first the variety of education's subcategories:</center></font>

In [10]:
print(df['education'].unique())

["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']


<font size='4'><center>lets lowercase everything and change 'some college' to 'college'</center></font>

In [11]:
df['education'] = df['education'].str.lower()
df['education'] = df['education'].replace('some college', 'college')
df.education.value_counts()

secondary education    15233
bachelor's degree       5260
college                  744
primary education        282
graduate degree            6
Name: education, dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Well done! :) 
</div>



<font size='4'><center>now education's fine</center></font>

### looking into the gender column

In [12]:
print(df.gender.value_counts())

F      14236
M       7288
XNA        1
Name: gender, dtype: int64


<font size = '4'><center>theres one weird instance of a third gender. lets check this row:</center></font>

In [13]:
print(df.loc[df['gender'] == 'XNA'])

       children  dob_years education  education_id      family_status  \
10701         0         24   college             2  civil partnership   

       family_status_id gender income_type  debt  total_income  \
10701                 1    XNA    business     0     32624.825   

               purpose  
10701  buy real estate  


<font size = '4'><center>the row seems legit, as XNA might be a technical value that some function generates on weird cases, like this.. 
<div class="alert alert-block alert-info">
<b>But we can't be sure. its only 1 row out of 20000, so in this case nothing will happen if we'll delete it:</b> <a class="tocSkip"></a>
</div></center></font>

In [14]:
df.drop(df.index[10701], inplace=True)

print(df.gender.value_counts())

F    14236
M     7288
Name: gender, dtype: int64


### Conclusion

<font size='4'><center>we've decided to drop the days_employed column due to unreasonable data</center></font>
    
<div class="alert alert-block alert-info">
<b>In addition, we deleted 1 row due to suspicious data (genter of type XNA).</b> <a class="tocSkip"></a>
</div>

<font size='4'><center>besides that, we've gone through all the columns manually and performed column.value_counts() on each.
no further action needs to be taken. only that it seems there is an option to mention youre unemployed, which
makes our explanation of missing 'total_income' from before a bit harder to convince by. but we'll stick with it.
people might feel discomfort to say theyre unemployed, and might not see the option when filling the form.
</center></font>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
~Unfortunately, XNA is an error in this column.It is only 1 row and we can delete it. :)~</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Good decision! </div>

## Data preprocessing

### Processing missing values

<font size='4'><center>now lets fill 'total_income' missing values. they came together with missing 'days_employed' values so the leading theory is that its mostly people who are currently unemployed. therefore, at present, they dont have any total_income at all. otherwise, we can argue its not possible for a business or a civil servant not to have
any income at all. in that case, we'll take the mean of 'total_income' for business and civil servants seperatly.
i.e - df.groupby('income_type')['total_income'].mean().
but, as said, we'll go with filling them with 0's:</center></font>

<div class="alert alert-block alert-info">
<b>OH WAIT!!! CHANGE OF PLAN!!
we'll fill the missing values with the medians for the appropriate income_type. turns out its not so simple as mentioned in the paragraph above. after trying a LOT of code i discovered this transform method code in the project's thread at slack..
</b> <a class="tocSkip"></a>
</div>

In [15]:
df['total_income'] = df.groupby('income_type')['total_income'].transform(lambda x: x.fillna(x.median()))

<font size='4'><center>using the method fillna() as this is the method we learned for tackling that task</center></font>
<font size='4'><center>lets check the filling's proper:</center></font>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

~Look, we have different proffecions here. And all of them have the different median income.~
    
~It would be more real to filling missing numbers by grouping on income_type and use median for different income. :)~ 
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Excellent! </div>

In [16]:
print(df.loc[df['total_income'].isna()]['income_type'].count())

0


<font size='5'><center>lets check for hiding 'None's:</center></font>

In [17]:
columns = df.columns
lst = []
for column in columns:
    tmp = df.loc[df[column] == 'None',column].count()
    lst.append(tmp)
lst

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

<font size='4'><center>we got a tuple of 11 0's, meaning theres no 'None' hiding in the data</center></font>

### Conclusion

<font size ='4'><center>we've conclude the data is free of missing values, including 'None' strings</center></font>

## Data type replacement

### changing 'total_income' to ints
<font size='4'><center>as there is no need for a float precision. all the other ints are 64 so we'll
make it 64 also</center></font>

In [18]:
df.loc[:,'total_income'] = df.loc[:,'total_income'].astype('int64')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21524 non-null  int64 
 1   dob_years         21524 non-null  int64 
 2   education         21524 non-null  object
 3   education_id      21524 non-null  int64 
 4   family_status     21524 non-null  object
 5   family_status_id  21524 non-null  int64 
 6   gender            21524 non-null  object
 7   income_type       21524 non-null  object
 8   debt              21524 non-null  int64 
 9   total_income      21524 non-null  int64 
 10  purpose           21524 non-null  object
dtypes: int64(6), object(5)
memory usage: 2.0+ MB


<font size='4'><center>using the astype() (as this is the method we learned for tackling that task) we got just 2 data
types in our table - object and int64</center></font>
<div class="alert alert-block alert-info">
<b>by calling the info method we see the table consists of 2 data types - object and int64</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Very good!
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Also we can check all data types. :) 
</div>

### Conclusion

### Processing duplicates

<font size ='5'><center>lets check for duplicated rows:</center></font>

In [19]:
df.duplicated().sum()

71

<font size = '4'><center>we found 71 of them, so lets drop them</center></font>

In [20]:
df = df.drop_duplicates().reset_index(drop=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21453 non-null  int64 
 1   dob_years         21453 non-null  int64 
 2   education         21453 non-null  object
 3   education_id      21453 non-null  int64 
 4   family_status     21453 non-null  object
 5   family_status_id  21453 non-null  int64 
 6   gender            21453 non-null  object
 7   income_type       21453 non-null  object
 8   debt              21453 non-null  int64 
 9   total_income      21453 non-null  int64 
 10  purpose           21453 non-null  object
dtypes: int64(6), object(5)
memory usage: 1.8+ MB


<font size = '4'><center>21524 - 21453 = 71</center></font>
<font size = '4'><center>so duplicates droped verified!
ive used the methods above for handling duplicates as this is the methods we learned for tackling that task
the duplicated data might have been present due to loan seekers applying twice, perhaps after closing the 
confirmation for sending their details, and then filling the form again to make sure their apply reach its destiny.
54 duplicates in our data is one duplicate for every 400 applies for loan. so our explanation above or some other
random reason are completely possible</center></font>

### Conclusion

<font size='4'><center>we found and removed all the duplicates in the data</center></font>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Great, duplicates are error and we do not need them. :) 
</div>

### Categorizing Data

#### categorizing the purpose column


In [21]:
def purpose_group(purpose): #function for categorizing the purpose column
    if 'real estate' in purpose or 'house' in purpose or 'housing' in purpose or 'property' in purpose:
        return 'real estate'
    if 'education' in purpose or 'university' in purpose or 'educated' in purpose:
        return 'education'
    if 'wedding' in purpose:
        return 'wedding'
    if 'car' in purpose:
        return 'car'
    else:
        return 'unknown'
    
df['purpose_type'] = df['purpose'].apply(purpose_group)
df['purpose_type'].value_counts()

#we see a lot of 'unknown's. lets look at their respective original purpose:

print(df.loc[df['purpose_type'] == 'unknown']['purpose'].value_counts())

#by tweeking the purpose_group function and rerunning the line above we categorise all the purposes until
#nothing gets printed. cant think of a better way. adding 'property' and 'house' to 'real estate' category
#needs to be done manually. nltk can be used specifically to categorise 'educated' as 'education', in
#this case its not worth the bother, especially if your not sharp in lemmatising and all yet.

Series([], Name: purpose, dtype: int64)


<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

~Why housing and real estate are different purposes?~
 
</div>

<div class="alert alert-block alert-info">
<b>now housing counts as real-estate</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
We can say that housin and real estate is different, but we need an explanation. For explane:

- under construction or old building
- commercial or noncommercial place
- repair or byu.

We have many variantы.Also, we can say that both are real estate. :) </div>

#### categorizing the 'children' column

In [22]:
print(df.children.value_counts())

 0     14090
 1      4808
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


<div font size='5'><center>we get strange children values: 20 and -1. because we dont get other strange values, but just those two, we'll
figure its something connected specifically to those numbers.
-1 might mean negation, meaning no children. therefore we'll replace it with 0.
right after typing a number using the keyboard's right side, you might hit the key below it and make it get
typed.
if you typed 1 or 3, youll notice something strange (right arrow or .) gets typed, that might generate an error
when submitting a loan request.
but, if you typed 2 youll end up with 20 and the loan request form could be sent with no problem. so we'll replace
20 with 2, and -1 with 0</center></div>

In [23]:
df['children'] = df['children'].replace({-1: 1, 20: 2})
print(df.children.value_counts())

0    14090
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64


<font size = '4'><center>children column done!</center></font>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Also we can say: -1 = 1. :) 
 
</div>

<div class="alert alert-block alert-info">
<b>now -1 childs = 1 child:)</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
OK. :) </div>

#### categorizing the 'total_income' column

In [24]:
max_income = df.loc[df['total_income'].idxmax()]['total_income']
min_income = df.loc[df['total_income'].idxmin()]['total_income']
mean_income = df['total_income'].mean()
median_income  = df.total_income.median()
print(max_income,min_income,mean_income,median_income)

362496 3306 26450.314920990073 22815.0


<font size = '4'><center>from those values we'll split the total_income data to 4 categories: 2 for the half that earns below the median,
and another 2 categories for the other half who earns more than the median</center></font>

In [25]:
def income_categorizer(income):
    if income < 11400:
        return 1
    if income < 22800:
        return 2
    if income < 200000:
        return 3
    else:
        return 4
df['total_income_id'] = df['total_income'].apply(income_categorizer)

### Conclusion

<div class="alert alert-block alert-info">
<b><font size = '4'><center>the data is categorised. we've categorized purpose into 4 categories, fixed problematic values in the children column, and categorised total_income into 4 categories - 2 for under the median and 2 for above the median</center></font></b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

It would be informative to add words: high, low, median, etc. :) 
 
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Good idea! :) </div>

## Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [26]:
debt_by_child = df.groupby('children').agg({'debt': ['sum', 'count']})
debt_by_child['ratio'] = debt_by_child['debt']['sum'] / debt_by_child['debt']['count']
debt_by_child.sort_values(by='ratio', ascending=False, inplace=True)
print(debt_by_child)

          debt            ratio
           sum  count          
children                       
4            4     41  0.097561
2          202   2128  0.094925
1          445   4855  0.091658
3           27    330  0.081818
0         1063  14090  0.075444
5            0      9  0.000000


### Conclusion

<font size = '4'><center>we see that people with 5 children dont go in debt. theres not a lot of them, but still, until we here otherwise,
we count on them when giving loans. maybe at their stage their kids help with the home chores more than "normal" kids,
so they can work more.
people with no children or 3 children have around 8% debt rate, and the others have roughly 20% chance more to get in debt.
</center></font>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Intresting differece! </div>

- Is there a relation between marital status and repaying a loan on time?

In [27]:
debt_by_family = df.groupby('family_status').agg({'debt': ['sum', 'count']})
debt_by_family['ratio'] = debt_by_family['debt']['sum'] / debt_by_family['debt']['count']
debt_by_family.sort_values(by='ratio', ascending=False, inplace=True)
print(debt_by_family)

                  debt            ratio
                   sum  count          
family_status                          
unmarried          274   2810  0.097509
civil partnership  388   4150  0.093494
married            931  12339  0.075452
divorced            85   1195  0.071130
widow / widower     63    959  0.065693


### Conclusion

<font size = '4'><center>we see widowers are less likely to get into debt, with 6.5% debt rate. on the other hand, unmarried and civil partnerships have 150% greater chance to get in debt</center></font>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Little surprise here. :) 
</div>

- Is there a relation between income level and repaying a loan on time?

In [28]:
debt_by_income_id = df.groupby('total_income_id').agg({'debt': ['sum', 'count']})
debt_by_income_id['ratio'] = debt_by_income_id['debt']['sum'] / debt_by_income_id['debt']['count']
debt_by_income_id.sort_values(by='ratio', ascending=False, inplace=True)
print(debt_by_income_id)

                debt            ratio
                 sum  count          
total_income_id                      
4                  1     11  0.090909
2                701   8187  0.085624
3                930  11681  0.079616
1                109   1574  0.069250


### Conclusion

<font size = '4'><center>its pretty surprising that the people who earn the most actually have the heighest debt rate. they might take much bigger lones than others, and therefore raise the chance of getting into debt. we'll recommend notice the loan details when dealing with high income people. altough its good for us that customers get into debt. especially if theyre rich. and anyway we should notice we only have 11 people who have total_income_id of 4, and only one of them is in debt, so take the data with caution.
the lowest earning sector has the lowest debt rate, at 6.9%. 
</center></font>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Great! Our results are not expected.  :)


</div>

- How do different loan purposes affect on-time repayment of the loan?

In [29]:
debt_by_purpose = df.groupby('purpose_type').agg({'debt':['sum', 'count']})
debt_by_purpose['ratio'] = debt_by_purpose['debt']['sum'] / debt_by_purpose['debt']['count']
debt_by_purpose.sort_values(by='ratio', ascending=False, inplace=True)
print(debt_by_purpose)

             debt            ratio
              sum  count          
purpose_type                      
car           403   4306  0.093590
education     370   4013  0.092200
wedding       186   2324  0.080034
real estate   782  10810  0.072340


### Conclusion

<font size = '4'><center>real-estate with the best debt rate at 7.2%, and car and education related
loans with 130% bigger chance of getting in debt, with around 9.3% debt rate for each
</center></font>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Yes, you are right about housing loans. They have more detailed checks. Perhaps, it help to avoid debts. :)
</div>

## General conclusion

<font size = '4'><center>the ideal lone taker is a widow/er loner with 5 or 0 kids (also 3 childs is ok), thats not earning way too much (actually, not at all much at less than half the mean income), and needs the lone for real estate related issues</center></font>
<div class="alert alert-block alert-info">
<b>the ideal loner has around 2/3 chance to get in debt compared to the average customer</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
In the conclusion should describe whole project here. We knew intresting information. It is very good to write details at the end. :) 
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Thank you! :) </div>

## Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [ ]  file examined;
- [ ]  missing values defined;
- [ ]  missing values are filled;
- [ ]  an explanation of which missing value types were detected;
- [ ]  explanation for the possible causes of missing values;
- [ ]  an explanation of how the blanks are filled;
- [ ]  replaced the real data type with an integer;
- [ ]  an explanation of which method is used to change the data type and why;
- [ ]  duplicates deleted;
- [ ]  an explanation of which method is used to find and remove duplicates;
- [ ]  description of the possible reasons for the appearance of duplicates in the data;
- [ ]  data is categorized;
- [ ]  an explanation of the principle of data categorization;
- [ ]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [ ]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [ ]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [ ]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [ ]  conclusions are present on each stage;
- [ ]  a general conclusion is made.